File: patch_step9.js
#!/usr/bin/env node
// Patch diagnose.js step9 to add LLM bullet generation
var fs = require('fs');
var path = 'skills/amazon-listing-doctor/diagnose.js';
var f = fs.readFileSync(path, 'utf8');

// Find the step9 block to replace
var searchStr = 'async function step9(ctx) {'
  + '\n  var s2 = ctx.deps[2].data;'
  + '\n  var s11 = ctx.deps[11] ? ctx.deps[11].data : null;'
  + '\n  var s12 = ctx.deps[12] ? ctx.deps[12].data : null;'
  + '\n  var bullets = s2.bullets || [];'
  + '\n  var violations = s11 ? (s11.violations || []) : [];'
  + '\n  var missingEgeo = s12 ? (s12.egeoFeatures || []).filter(function(f) { return f.missing; }) : [];'
  + '\n\n  var violatedBullets = {};'
  + '\n  violations.forEach(function(v) {'
  + '\n    if (v.bullet) violatedBullets[v.bullet - 1] = v;'
  + '\n  });'
  + '\n\n  var optimized = bullets.map(function(bulletText, i) {'
  + '\n    var violationActions = [];'
  + '\n    var enhanceActions = [];'
  + '\n    var text = bulletText;';

var replaceStr = 'async function step9(ctx) {'
  + '\n  var s2 = ctx.deps[2].data;'
  + '\n  var s11 = ctx.deps[11] ? ctx.deps[11].data : null;'
  + '\n  var s12 = ctx.deps[12] ? ctx.deps[12].data : null;'
  + '\n  var bullets = s2.bullets || [];'
  + '\n  var violations = (s11 ? (s11.violations || []) : []).filter(function(v) { return v.severity && v.severity !== \'none\'; });'
  + '\n  var missingEgeo = s12 ? (s12.egeoFeatures || []).filter(function(f) { return f.missing; }) : [];'
  + '\n\n  // ── LLM-driven bullet generation (step9) ──'
  + '\n  var stepLLM = require(\'./stepLLM\');'
  + '\n  log(\'  [step9] generating optimized bullets via LLM...\');'
  + '\n  var llmBullets = null;'
  + '\n  try {'
  + '\n    llmBullets = await stepLLM.generateOptimizedBullets({'
  + '\n      bullets: bullets,'
  + '\n      violations: violations,'
  + '\n      implicitViolations: s11 ? (s11.implicitViolations || []) : [],'
  + '\n      missingEgeo: missingEgeo'
  + '\n    });'
  + '\n  } catch(e) {'
  + '\n    log(\'  [step9] LLM bullet gen failed: \' + e.message + \' — falling back to template\');'
  + '\n  }'
  + '\n  if (llmBullets && llmBullets.bullets && llmBullets.bullets.length > 0) {'
  + '\n    var bulletMap = {};'
  + '\n    llmBullets.bullets.forEach(function(b) {'
  + '\n      if (b.index !== undefined) bulletMap[b.index - 1] = b;'
  + '\n    });'
  + '\n    var optimized = bullets.map(function(bulletText, i) {'
  + '\n      if (bulletMap[i]) {'
  + '\n        var b = bulletMap[i];'
  + '\n        var parts = [b.rewritten || bulletText];'
  + '\n        if (b.actions && b.actions.length > 0) {'
  + '\n          b.actions.forEach(function(a) { parts.push(\'  \\u2192 \' + a); });'
  + '\n        }'
  + '\n        return parts.join(\'\\n\');'
  + '\n      }'
  + '\n      return bulletText;'
  + '\n    });'
  + '\n    return { optimized: optimized, missingEgeo: missingEgeo.map(function(f) { return f.label; }), method: \'llm\' };'
  + '\n  }'
  + '\n  // ── Fallback: Template-based bullet optimization ──'
  + '\n  var violatedBullets = {};'
  + '\n  violations.forEach(function(v) {'
  + '\n    if (v.bullet) violatedBullets[v.bullet - 1] = v;'
  + '\n  });'
  + '\n  var optimized = bullets.map(function(bulletText, i) {'
  + '\n    var violationActions = [];'
  + '\n    var enhanceActions = [];'
  + '\n    var text = bulletText;';

var idx = f.indexOf(searchStr);
if (idx < 0) {
  console.log('Could not find step9 search string');
  process.exit(1);
}
console.log('Found step9 at char', idx);
var newF = f.slice(0, idx) + replaceStr + f.slice(idx + searchStr.length);
fs.writeFileSync(path, newF, 'utf8');
console.log('Patched successfully');

File: patch_step9.py
#!/usr/bin/env python3
import sys

path = r'C:\Users\csbd\.openclaw\workspace\e-commerce-tools\skills\amazon-listing-doctor\diagnose.js'
with open(path, 'rb') as f:
    data = f.read()

CRLF = b'\r\n'

search = (
    CRLF + b'  var violations = s11 ? (s11.violations || []) : [];' + CRLF +
    b'  var missingEgeo = s12 ? (s12.egeoFeatures || []).filter(function(f) { return f.missing; }) : [];' + CRLF +
    CRLF + CRLF +
    b'  var violatedBullets = {};' + CRLF +
    b'  violations.forEach(function(v) {' + CRLF +
    b'    if (v.bullet) violatedBullets[v.bullet - 1] = v;' + CRLF +
    b'  });' + CRLF +
    CRLF + CRLF +
    b'  var optimized = bullets.map(function(bulletText, i) {' + CRLF +
    b'    var violationActions = [];' + CRLF +
    b'    var enhanceActions = [];' + CRLF +
    b'    var text = bulletText;'
)

replace = (
    CRLF + b"  var violations = (s11 ? (s11.violations || []) : []).filter(function(v) { return v.severity && v.severity !== 'none'; });" + CRLF +
    b"  var missingEgeo = s12 ? (s12.egeoFeatures || []).filter(function(f) { return f.missing; }) : [];" + CRLF +
    CRLF +
    b"  // LLM-driven bullet generation (step9)" + CRLF +
    b"  var stepLLM = require('./stepLLM');" + CRLF +
    b"  log('  [step9] generating optimized bullets via LLM...');" + CRLF +
    b"  var llmBullets = null;" + CRLF +
    b"  try {" + CRLF +
    b"    llmBullets = await stepLLM.generateOptimizedBullets({" + CRLF +
    b"      bullets: bullets," + CRLF +
    b"      violations: violations," + CRLF +
    b"      implicitViolations: s11 ? (s11.implicitViolations || []) : []," + CRLF +
    b"      missingEgeo: missingEgeo" + CRLF +
    b"    });" + CRLF +
    b"  } catch(e) {" + CRLF +
    b"    log('  [step9] LLM bullet gen failed: ' + e.message + ' falling back to template');" + CRLF +
    b"  }" + CRLF +
    b"  if (llmBullets && llmBullets.bullets && llmBullets.bullets.length > 0) {" + CRLF +
    b"    var bulletMap = {};" + CRLF +
    b"    llmBullets.bullets.forEach(function(b) {" + CRLF +
    b"      if (b.index !== undefined) bulletMap[b.index - 1] = b;" + CRLF +
    b"    });" + CRLF +
    b"    var optimized = bullets.map(function(bulletText, i) {" + CRLF +
    b"      if (bulletMap[i]) {" + CRLF +
    b"        var b = bulletMap[i];" + CRLF +
    b"        var parts = [b.rewritten || bulletText];" + CRLF +
    b"        if (b.actions && b.actions.length > 0) {" + CRLF +
    b"          b.actions.forEach(function(a) { parts.push('  -> ' + a); });" + CRLF +
    b"        }" + CRLF +
    b"        return parts.join('\\n');" + CRLF +
    b"      }" + CRLF +
    b"      return bulletText;" + CRLF +
    b"    });" + CRLF +
    b"    return { optimized: optimized, missingEgeo: missingEgeo.map(function(f) { return f.label; }), method: 'llm' };" + CRLF +
    b"  }" + CRLF +
    b"  // Fallback: Template-based bullet optimization" + CRLF +
    b"  var violatedBullets = {};" + CRLF +
    b"  violations.forEach(function(v) {" + CRLF +
    b"    if (v.bullet) violatedBullets[v.bullet - 1] = v;" + CRLF +
    b"  });" + CRLF +
    CRLF +
    b"  var optimized = bullets.map(function(bulletText, i) {" + CRLF +
    b"    var violationActions = [];" + CRLF +
    b"    var enhanceActions = [];" + CRLF +
    b"    var text = bulletText;"
)

idx = data.find(search)
if idx < 0:
    print('NOT FOUND')
    sys.exit(1)

print('Found at byte', idx)
new_data = data[:idx] + replace + data[idx+len(search):]
with open(path, 'wb') as f:
    f.write(new_data)
print('Patched OK, new len:', len(new_data))
